File: executor/src/main/scala/com/oceanum/pluggable/PluggableExecutor.scala
package com.oceanum.pluggable

import akka.actor.{ActorRef, Props}

abstract class PluggableExecutor {
  private lazy val endpoint: ActorRef = system.actorOf(Props(classOf[PluggableSubEndPoint], primEndpoint, this))

  def updateInfo(info: PluggableInfo): Unit = endpoint ! info

  def updateState(state: PluggableState): Unit = endpoint ! state

  def startRun(): TaskComplete

  def kill(): Unit

  def close(): Unit = {
    system.terminate()
    System.exit(0)
  }
}

File: executor/src/main/java/com/oceanum/pluggable/javadsl/PluggableExecutor.java
package com.oceanum.pluggable.javadsl;

import com.oceanum.pluggable.ExecutionProxy;
import com.oceanum.pluggable.TaskComplete;

import java.util.Map;
import java.util.function.Function;

public abstract class PluggableExecutor {
    private ExecutionProxy proxy;

    public void setProxy(Function<PluggableExecutor, ExecutionProxy> proxyFunc) {
        this.proxy = proxyFunc.apply(this);
    }

    public void updateInfo(PluggableInfo info) {
        proxy.updateInfo(info);
    }

    public void updateState(PluggableState state) {
        proxy.updateState(state);
    }

    public abstract TaskComplete startRun(Map<String, Object> args);

    public abstract boolean kill();

    public abstract boolean isKilled();
}
